Add profile support for jit-compiled functions.

stefanseefeld opened the pull request:
This patch allows profilers to see jit-compiled functions.
The patch creates a synthetic frame object and attaches the code object from the original (python) function to it, so the functions shows up in the profiler with the location of the non-compiled python code.
The one thing still missing is a 'locals' dictionary containing the objects referenced as function arguments. I'm not sure yet how to efficiently create that.

pitrou commented on numba/_dispatcher.c:
```diff
@@ -6,6 +6,92 @@
+           PyThreadState *tstate, PyFrameObject *frame,
+           int what, PyObject *arg)
+{
+  int result;
```
Can you set your editor to 4-spaces indent for C code?

pitrou commented on numba/_dispatcher.c:
```diff
@@ -6,6 +6,83 @@
+    if (err == 0)
+    {
+        PyErr_Restore(type, value, traceback);
+	return 0;
```
Can you ensure your editor doesn't emit tabs? :-)

pitrou commented on numba/_dispatcher.c:
```diff
@@ -6,6 +6,83 @@
 #include "_typeof.h"
+#include "frameobject.h"
+
+#define PROFILE_CALL 1
```
I'm not sure why a macro is necessary, but if it is, you should add a one-line description.

stefanseefeld replied:
Right, I will remove it. (It was a simple means to switch between the "old" and the "new" code.)

pitrou commented on numba/_dispatcher.c:
```diff
@@ -6,6 +6,83 @@
+
+#if PROFILE_CALL
+
+static int
```
Explain that the following functions and macros are taken from ceval.c perhaps?

stefanseefeld replied:
Will do.

pitrou commented on numba/_dispatcher.c:
```diff
@@ -191,18 +268,53 @@ int search_new_conversions(PyObject *dispatcher, PyObject *args, PyObject *kws)
+        // now populate the 'fast locals' in `frame`
+        frame->f_locals = locals;
+        PyFrame_LocalsToFast(frame, 0);
+        frame->f_locals = NULL;
```
Why this?

stefanseefeld replied:
I don't know, TBH. The whole PyFrame_LocalsToFast and PyFramr_FastToLocals business isn't well documented, and I was trying to figure out which members are expected to be valid and which aren't.

pitrou replied:
Well, if you set `f_locals` to NULL, the frame destructor won't decref the locals, which is why you have to decref them yourself :-)

stefanseefeld replied:
The original locals is a `borrowed reference` (from a call to PyEval_GetLocals()), so I didn't think it had to be decrefed in the first place. Does PyFrame_LocalsToFast() increment the reference to it ?

pitrou replied:
You seem to be right. OTOH, the frame is supposed to own the reference to f_locals, which means you would actually have to incref those locals (and, before, decref the previous f_locals value).

pitrou commented on numba/_dispatcher.c:
```diff
@@ -191,18 +268,53 @@ int search_new_conversions(PyObject *dispatcher, PyObject *args, PyObject *kws)
+    if (tstate->use_tracing && tstate->c_profilefunc)
+    {
+        /*
+	 * The following code requires some explaining:
```
You should explain what you create a dummy frame at all :-)

pitrou commented on numba/_dispatcher.c:
```diff
@@ -191,18 +268,53 @@ int search_new_conversions(PyObject *dispatcher, PyObject *args, PyObject *kws)
+	 * member, then call `PyFrame_LocalsToFast`, where a subsequent call to the `frame.f_locals`
+	 * property (by virtue of the `frame_getlocals` function in frameobject.c) will find them.
+	 */
+        PyCodeObject *code = (PyCodeObject*)PyObject_GetAttrString((PyObject*)self, "func_code");
```
`func_code` doesn't exist in Python 3, you want `__code__` instead.

pitrou replied:
Also, you probably want to check for errors here...

stefanseefeld replied:
OK. (The _DispatcherBase constructor actually sets both, so this shouldn't matter.)

pitrou commented on numba/_dispatcher.c:
```diff
@@ -191,18 +268,53 @@ int search_new_conversions(PyObject *dispatcher, PyObject *args, PyObject *kws)
+        frame->f_locals = locals;
+        PyFrame_LocalsToFast(frame, 0);
+        frame->f_locals = NULL;
+        PyObject *result;
```
C89 doesn't allow intermingled declarations and statements (MSVC will choke over this).

pitrou commented on numba/tests/test_profile_refcount.py:
```diff
@@ -0,0 +1,34 @@
+from __future__ import division, absolute_import, print_function
+
+import numba.unittest_support as unittest
```
Just a nit, but can you follow PEP 8 for imports order? (stdlib then third-party then current project)
It makes it easier to figure out a module's dependencies.
See https://www.python.org/dev/peps/pep-0008/#imports

pitrou commented on numba/tests/test_profile_refcount.py:
```diff
@@ -0,0 +1,34 @@
+class TestProfileRefCount(MemoryLeakMixin, unittest.TestCase):
+
+    def assert_refcount(self, arr, expect):
+        self.assertEqual(arr.base.refcount, expect)
```
I didn't know `.refcount` and I don't think it's necessary to go through this. Just compare `sys.getrefcount(arr)` before and after the function call, IMO.

pitrou commented on numba/tests/test_profile_refcount.py:
```diff
@@ -0,0 +1,34 @@
+import sys
+
+
+class TestProfileRefCount(MemoryLeakMixin, unittest.TestCase):
```
Perhaps add a comment or docstring why these tests are useful?

stefanseefeld replied:
After I modified the patch to pass 'locals' as a function argument rather than keep it as a Dispatcher member this test didn't fail any longer, so it isn't good enough to measure correctness, unfortunately. I'm still chasing for a) the real issue and b) a good indicator for the issue's presence.

pitrou commented on numba/tests/test_profile_refcount.py:
```diff
@@ -0,0 +1,34 @@
+        self.assertEqual(initrefct + 1, sys.getrefcount(arr))
+        del got
+
+        p.disable()
```
You should ensure this is always called, using e.g. `self.addCleanup`.

stefanseefeld commented:
I have reworked the patch a bit, and added an incref to the locals when they are passed to the frame object. (I have to admit that I find the entire handling of locals using the frame API confusing, or even wrong (The PyFrame_New() constructor takes a locals argument, but then ignores it. That can't be right !).
At least now the code is consistent. And the tests still pass. :-)
So I think this is ready to be merged.

pitrou commented:
Ok, so the refleak has disappeared?
How about adding some simple test that the function does show up in the profiler's output?
(also, be sure to fix the indentation of the C code :-))

pitrou commented on numba/_dispatcher.c:
```diff
@@ -191,18 +269,52 @@ int search_new_conversions(PyObject *dispatcher, PyObject *args, PyObject *kws)
+        PyFrameObject *frame = PyFrame_New(tstate, code, globals, NULL);
+        PyObject *result;
+        // now populate the 'fast locals' in `frame`
+        if (frame->f_locals)
```
Py_XDECREF already does the NULL check for you.

pitrou commented on numba/_dispatcher.c:
```diff
@@ -191,18 +269,52 @@ int search_new_conversions(PyObject *dispatcher, PyObject *args, PyObject *kws)
+         * member, then call `PyFrame_LocalsToFast`, where a subsequent call to the `frame.f_locals`
+         * property (by virtue of the `frame_getlocals` function in frameobject.c) will find them.
+         */
+        PyCodeObject *code = (PyCodeObject*)PyObject_GetAttrString((PyObject*)self, "__code__");
```
It would be good to check that `code` isn't NULL here.

stefanseefeld commented:
The ref leak has disappeared when I didn't reset frame->f_locals to NULL, as you suggested. I then added an incref for symmetry (to make the code *look* correct), which didn't seem to make an observable difference.
And I have corrected the indentation (including untabifying the whole file).

pitrou commented on numba/_dispatcher.c:
```diff
@@ -191,18 +269,56 @@ int search_new_conversions(PyObject *dispatcher, PyObject *args, PyObject *kws)
+            PyErr_Format(PyExc_RuntimeError, "No __code__ attribute found.");
+            return NULL;
+        }
+        PyObject *globals = PyDict_New();
```
Those variables must be declared before any other statement in the block.

pitrou commented:
Ok, good enough :) Thank you!
